src/lib/unlang/xlat_builtin.c

```diff
@@ -1317,47 +1317,54 @@ static xlat_action_t xlat_func_next_time(TALLOC_CTX *ctx, fr_dcursor_t *out, req
 	return XLAT_ACTION_DONE;
 }
 
+static xlat_arg_parser_t const xlat_func_xlat_args[] = {
+	{ .required = true, .concat = true, .type = FR_TYPE_STRING },
+	XLAT_ARG_PARSER_TERMINATOR
+};
+
 /** xlat expand string attribute value
  *
  * @ingroup xlat_functions
  */
-static ssize_t xlat_func_xlat(TALLOC_CTX *ctx, char **out, size_t outlen,
-			      UNUSED void const *mod_inst, UNUSED void const *xlat_inst,
-			      request_t *request, char const *fmt)
+static xlat_action_t xlat_func_xlat(TALLOC_CTX *ctx, fr_dcursor_t *out, request_t *request,
+				    UNUSED void const *xlat_inst, UNUSED void *xlat_thread_inst,
+				    fr_value_box_list_t *in)
 {
 	ssize_t		slen;
 	fr_pair_t	*vp;
+	fr_value_box_t	*in_head = fr_dlist_head(in);
+	fr_value_box_t	*vb;
 
-	fr_skip_whitespace(fmt);
-
-	if (outlen < 3) {
-	nothing:
-		return 0;
-	}
-
-	if ((xlat_fmt_get_vp(&vp, request, fmt) < 0) || !vp) goto nothing;
+	if ((xlat_fmt_get_vp(&vp, request, in_head->vb_strvalue) < 0) || !vp) return XLAT_ACTION_FAIL;
 
-	RDEBUG2("EXPAND %s", fmt);
+	RDEBUG2("EXPAND %pV", in_head);
 	RINDENT();
 
+	MEM(vb = fr_value_box_alloc_null(ctx));
 	/*
 	 *	If it's a string, expand it again
 	 */
 	if (vp->vp_type == FR_TYPE_STRING) {
-		slen = xlat_eval(*out, outlen, request, vp->vp_strvalue, NULL, NULL);
-		if (slen <= 0) return slen;
+		char	buff[1024];
+		slen = xlat_eval(buff, 1024, request, vp->vp_strvalue, NULL, NULL);
```

Comment:
Not the correct function to use here.  You need to use the async capable call.  You might just be able to get away with parenting the expansion off the ctx returned by `interpreter_frame_talloc_ctx()`, and pushing the new xlat onto the interpreter stack for evaluation.